Add On-Chain Send/Receive to NWC

barrydeen opened the pull request:
- Introduces two new methods: `pay_chain_address` and `make_chain_address`
 - Modifies `list_transactions` to include an optional field `txid`

### `pay_chain_address`

Description: Requests payment to a chain address.

Request:

```jsonc
{
  "method": "pay_chain_address",
  "params": {
    "address": "bc1q...", // chain address
    "amount": 12000 // amount in sats
  }
}
```

Response:

```jsonc
{
  "result_type": "pay_chain_address",
  "result": {
    "txid": "0123456789abcdef..." // chain tx id
  }
}
```

Errors:

- `PAYMENT_FAILED`: The payment failed.

### `make_chain_address`

Request:

```jsonc
{
  "method": "make_chain_address",
  "params": {
    "type": "p2tr" // address type, optional (p2tr, p2wphk)
  }
}
```

Response:

```jsonc
{
  "result_type": "make_chain_address",
  "result": {
    "address": "bc1q...", // chain address
    "type": "p2tr" // address type
  }
}
```

benthecarman commented:
A sign psbt and list utxos cmd would make sense as well. 

Also the balance command might want to include an onchain and unconfirmed balance now as well

barrydeen commented:
> A sign psbt and list utxos cmd would make sense as well.
> 
> Also the balance command might want to include an onchain and unconfirmed balance now as well

Agreed on all fronts.

barrydeen commented:
@benthecarman 

would you do sign psbt like this?

### `sign_psbt`

Request:

```jsonc
{
  "method": "sign_psbt",
  "params": {
    "psbt": "string" // base64 encoded psbt
  }
}
```

Response:

```jsonc
{
  "result_type": "sign_psbt",
  "result": {
    "psbt": "string" // base64 encoded signed psbt
  }
}
```

benthecarman commented:
Yeah, could add an optional input index too

supertestnet commented:
ACK

vitorpamplona commented:
Very cool!

But please don't change formatting with new ideas.

barrydeen commented:
> Very cool!
> 
> But please don't change formatting with new ideas.

Ah yes. Fixed.

barrydeen commented:
One thing I'm a bit unsure about is if chain amounts should be expressed as msats or sats. For now an msat can never settle on chain, but if you're building an app or wallet service I bet it will be extremely annoying to treat the amounts differently

rolznz commented on 47.md:
```diff
@@ -409,6 +487,7 @@ Response:
     "result_type": "get_balance",
     "result": {
         "balance": 10000, // user's balance in msats
+        "unconfirmed_balance": 10000, // optional, user's unconfirmed balance in msats
```
With this change we mix lightning and on-chain balances, which is a problem for all wallets that do not support a unified balance (most lightning wallets).

What about a `get_chain_balance`?

benthecarman replied:
yeah i think `get_chain_balance` makes more sense

barrydeen replied:
I suppose wallets that do unified can just return it from get_balance and not allow get_chain_balance as a method

barrydeen commented on 47.md:
```diff
@@ -377,6 +425,7 @@ Response:
             {
                "type": "incoming", // "incoming" for invoices, "outgoing" for payments
                "invoice": "string", // encoded invoice, optional
+               "txid": "string", // transaction id for on-chain, optional
```
yeah i guess we can do that and leave it up to the wallet interface if they unify them or not instead of the backend

janniks commented:
For signPsbt maybe take a look at [Unisats wallet API](https://docs.unisat.io/dev/unisat-developer-center/unisat-wallet#signpsbt). I believe it's the best one out there that's being used.
Leather wallet has done more research in terms of what can be done directly in PSBT info (fingerprints encoding which input should be signed), so technically a lot of the API isn't needed, but developers like the convenience.

We iterated and improved the available options to [this `signPsbt` spec](https://wbips.netlify.app/request_api/signPsbt) for WBIPs. And have 4 wallets using the standard or changing to it.

kumulynja commented:
Would it make sense to add an ALREADY_PAID error and recommend wallets to not pay the same amount to the same address twice?

This to reduce the risk of making the same payment twice in case the client for whatever reason didn't receive the response and retries a request? Or should this just be kept track of and be handled by any wallet how it seems fit and not really be mentioned in the specs?

barrydeen commented:
> For signPsbt maybe take a look at [Unisats wallet API](https://docs.unisat.io/dev/unisat-developer-center/unisat-wallet#signpsbt). I believe it's the best one out there that's being used. Leather wallet has done more research in terms of what can be done directly in PSBT info (fingerprints encoding which input should be signed), so technically a lot of the API isn't needed, but developers like the convenience.
> 
> We iterated and improved the available options to [this `signPsbt` spec](https://wbips.netlify.app/request_api/signPsbt) for WBIPs. And have 4 wallets using the standard or changing to it.

If you specify the input index, wouldn't the wallet service already know the pubkey and address?

supertestnet commented:
> I agree that it can be put in a failure reason and just use an internal or other error...

Still, an error message should be specified so that implementations don't diverge on this point. Also, I think there should be an override option. A user might *want* to send the same amount to the same address twice. E.g. what if it's a subscription service? Or what if a user decided to send 10k sats to an exchange once per week and sell it as a kind of "allowance"? If that exchange has a static deposit address, as some do, then if the wallet says "you're not supposed to send the same amount to the same address twice," the user should be able to override that and say "do it anyway."

benthecarman commented:
> Would it make sense to add an ALREADY_PAID error and recommend wallets to not pay the same amount to the same address twice?
> 
> This to reduce the risk of making the same payment twice in case the client for whatever reason didn't receive the response and retries a request? Or should this just be kept track of and be handled by any wallet how it seems fit and not really be mentioned in the specs?

I don't love this, you can already have idempotentcy from the event id, adding it redundantly like this just inhibits use cases and can be confusing

kumulynja commented:
> > Would it make sense to add an ALREADY_PAID error and recommend wallets to not pay the same amount to the same address twice?
> > This to reduce the risk of making the same payment twice in case the client for whatever reason didn't receive the response and retries a request? Or should this just be kept track of and be handled by any wallet how it seems fit and not really be mentioned in the specs?
> 
> I don't love this, you can already have idempotentcy from the event id, adding it redundantly like this just inhibits use cases and can be confusing

Totally agree, and that was where I was going to go, should we add a recommendation to the specs now that 'pay' retries should be done with the same event/event id? Because I can imagine naive client implementations may have their retry logic generate a completely new event...

kumulynja commented:
> > I agree that it can be put in a failure reason and just use an internal or other error...
> 
> Still, an error message should be specified so that implementations don't diverge on this point. Also, I think there should be an override option. A user might _want_ to send the same amount to the same address twice. E.g. what if it's a subscription service? Or what if a user decided to send 10k sats to an exchange once per week and sell it as a kind of "allowance"? If that exchange has a static deposit address, as some do, then if the wallet says "you're not supposed to send the same amount to the same address twice," the user should be able to override that and say "do it anyway."

Yes, I agree that should all be possible. I think the most straightforward way to handle it and permit those use cases is as @benthecarman mentions, wallet services should keep track of already processed events and not process them twice. 
Clients that have some retry mechanisms, where they try to get the same payment done in a short amount of time, should reuse the same event.

My worry is just that both wallet services as clients may not think about this initially, especially since the "risk" of double payments didn't exist with lightning invoices only. Since with on-chain payments it does exist, I thought we could at least add a recommendation on what to check in the wallet service and how to retry payments from the client side.

lnbc1QWFyb24 commented:
Stoked to see this @barrydeen. We will add this to the Clams Accounting app.

lnbc1QWFyb24 commented:
Oh yes, my bad, `list_chain_transactions` is what I meant 👍

benthecarman commented:
> Just highlighting I was on spaces with Cove wallet creator that was looking a way to connect a node to his mobile wallet and mentioned this may work with NWC. He's main request was "getting txn information for addresses"

That is completely separate and should be separate from nwc

barrydeen commented on 47.md:
````diff
@@ -274,6 +274,31 @@ pubkey should be used.
+Request:
+```jsonc
+{
+    "method": "pay_chain_address",
````
yes I think we should add this, i'll update the PR shortly. I think sat/vB is the way

f321x replied:
Just note that if only sat/vb is available clients **have** to fetch the feerate from some external source, while allowing for an eta would allow for simpler clients that don't have to do requests to some external source to get the current mempool state.

benthecarman replied:
Just make it optional then, the wallet will fetch it then

f321x commented on 47.md:
```diff
@@ -413,6 +524,27 @@ Response:
+{
+    "result_type": "get_chain_balance",
+    "result": {
+        "balance": 100000, // user's balance in sats
```
Does `balance` contain the unconfirmed balance? It might be better to explicitly name it `confirmed_balance` to make it less ambiguous

barrydeen commented:
Gentleman, major update, look forward to your comments:

- changed `pay_chain_address` to just `pay_chain` given that we can do multiple recipients
- `list_chain_transactions` now shows information about the spent and received utxos
- added `feerate` param to `pay_chain`

benthecarman reviewed (approved):
Lgtm

barrydeen commented:
> A method to get the current mempool state might be useful to have in this context, so the client can get fee estimates from its own backend and doesn't have to rely on other sources: e.g. `get_mempool_info`
> 
> ```json
> {
>     "method": "get_mempool_info",
>     "params": {}
> }
> ```
> 
> Response:
> 
> ```json
> {
>     "result_type": "get_mempool_info",
>     "result": {
>         "estimates": {    # blocks -> sat/vb
>           "1": "24",
>           "2": "20",
>           "5": "18",
>           "10": "10",
>           "25": "5",
>           "144": "2",
>           "1008": "1",
>       }
>     }
> }
> ```
> 
> the backend should be free to specify the available block values, the client shouldn't rely on these specific values. or some `NO_FEE_ESTIMATES_AVAILABLE_ERROR` in case no single value is available.

i just wonder how far we go down this route, in theory why not just expose the entire bitcoin rpc to nwc methods? (maybe we should!) - I don't disagree with your logic that it would make life easier for a client, I'm just not sure.

f321x commented on 47.md:
```diff
@@ -274,25 +274,28 @@ pubkey should be used.
 
-Description: Requests payment to a chain address.
+Description: Requests payment to one or more chain addresses.
 
```
What should happen if the backend is not able to pay all adresses but only some (e.g. too few funds)?
Might add a line saying something alike: `All given addresses have to be paid in a single transaction or the request gets failed`. 
This should give the client the explicit guarantee that the transaction will happen as requested.

f321x reviewed (approved):
looks reasonable, RE `get_mempool_info`, i agree there has to be some limitation on what calls really make sense for this usecase to include. 
I think some mempool information would be quite useful for clients that do onchain operations, and the optional nature of NIP-47 features would leave the decision open if the backend actually wants to implement it, otherwise they can just not signal support for it.

TheBlueMatt commented:
There was a bunch of discussion on using bip (3)21 as a standard way to encode flexible payment instructions in a future rev of nwc in #1952. Ultimately we decided to do a live call and I summarized some of our conclusions at https://github.com/nostr-protocol/nips/pull/1952#issuecomment-3739299802 ISTM that would be a simpler way to also get on-chain where we can design one thing that is forward-looking and flexible rather than trying to add new methods for every new payment method in the future. Of course some use-cases will need specific payment methods and may even need optional future extensions, but specifying it as generic as possible means those extensions can be quite simple and additional payment methods can even be totally transparent to some wallets.
